engine/src/main/java/com/datatorrent/stram/debug/TupleRecorderCollection.java

```diff
@@ -161,19 +161,13 @@ private void startRecording(String id, final Node<?> node, int operatorId, final
     if (!conflict) {
       logger.debug("Executing start recording request for {}", operatorIdPortNamePair);
 
-      if (gatewayAddress != null && wsClient == null) {
+      if ((wsClient != null) && !wsClient.isConnectionOpen()) {
         synchronized (this) {
-          if (wsClient == null) {
+          if ((wsClient != null) && !wsClient.isConnectionOpen()) {
```

Comment:
How can wsClient be `null` after check on line 164? What synchronized protects here? A connection can be lost asynchronously and synchronize does not guarantee that wsClient is connected immediately after `setup`.

Reply:
I was following your earlier suggestion for (wsClient != null) && !wsClient.isConnectionOpen(). The check in current codebase is that setup is not called twice, I guess the closest equivalent for it is just connection != null. That is not exposed so  I was using a state indicator in TupleRecorderCollection earlier. I guess we can add a isConnectionSetup method on PubSubWebSocketClient and use that in TRC instead. If for any reason connection doesn't get created for reasons such as the post failing (line 184 in PubSubWebSocketClient), it will be tried again when a newer recording is opened, which would be a good thing

Reply:
I don't see a need for `synchronized` as I don't understand what it protects. Without synchronize, there is no need for the first check, so check for `wsClient != null` is necessary. IMO, it should be the responsibility of PubSubWebSocketClient to handle connections and ensure that if `connection != null`, to reset or reuse it, so I'd prefer not to manage state outside of the Client or provide isConnectionSetup().

Reply:
What if there are two simultaneous tuple recording starts?